File: AI_Model_testing/ML/Computational_Intelligence/CI_Model/PSO/MainPSO.cpp
// MainPSO.cpp รับข้อมูลจาก Main_CL.cpp แล้วส่งไปคำนวน
#include "MainPSO.hpp"
#include "Core/Engine.hpp"

using namespace std;
namespace py = pybind11;

ResultRun run_bpso(py::array_t<double> input_X , py::array_t<int> input_y ,int num_classes) {

    auto buf = input_X.unchecked<2>();
    auto buf_y = input_y.unchecked<1>();

    int rows = buf.shape(0);
    int cols = buf.shape(1);

    vector<vector<double>> X(
        rows, vector<double>(cols)
    );
    vector<int> y(rows);

    for (int i = 0; i < rows; i++){
        y[i] = buf_y(i);
        for (int j = 0; j < cols; j++){
            X[i][j] = buf(i, j);
        }
    }
    int amount_particles = 0;
    int iterations = 10 * cols;

    if (cols < 20)
        amount_particles = 15;
    else if (cols <= 50)
        amount_particles = 25;
    else if (cols <= 100)
        amount_particles = 40;
    else if (cols <= 300)
        amount_particles = 60;
    else if (cols <= 600)
        amount_particles = 80;
    else if (cols <= 1000)
        amount_particles = 100;
    else
        amount_particles = 125;  

    unsigned int cpuCount = thread::hardware_concurrency();

    if (cpuCount == 0) cpuCount = 2; 
    else cpuCount -= 2;

    EnginePSO engine(amount_particles, iterations, X , y ,num_classes);
    EnginePSO::PSOResult res;

    if (amount_particles < 20)
    {
        res = engine.run_parallel(1);
    }
    else
    {
        res = engine.run_parallel(cpuCount);
    }

    ResultRun result;

    result.best_features = res.best_features;
    result.best_score = res.best_score;

    return result;
}

File: AI_Model_testing/ML/Computational_Intelligence/CI_Model/PSO/Python_PSO/MainPSO.py
# MainPSO.py - Python equivalent of MainPSO.cpp
# Prepares data matrix, sets particle count and iterations, runs EnginePSO

import numpy as np
from .Core.Engine import EnginePSO


def run_bpso(input_X, input_y, num_classes):
    """
    Exact 1:1 translation of C++ run_bpso in MainPSO.cpp
    Determines particle counts and iterations dynamically based on feature count.
    """
    if isinstance(input_X, np.ndarray):
        X = input_X.tolist()
    else:
        X = [list(row) for row in input_X]

    if isinstance(input_y, np.ndarray):
        y = input_y.astype(int).tolist()
    else:
        y = [int(v) for v in input_y]

    rows = len(X)
    cols = len(X[0])
    iterations = 10 * cols

    if cols < 20:
        amount_particles = 15
    elif cols <= 50:
        amount_particles = 25
    elif cols <= 100:
        amount_particles = 40
    elif cols <= 300:
        amount_particles = 60
    elif cols <= 600:
        amount_particles = 80
    elif cols <= 1000:
        amount_particles = 100
    else:
        amount_particles = 125

    engine = EnginePSO(amount_particles, iterations, X, y, num_classes)
    best_features, best_score = engine.run()

    return best_features, best_score
